SO_properties.py

```diff
@@ -636,12 +636,116 @@ def Tgas(self):
             return None
         return (self.gas_temperatures * (self.gas_masses / self.Mgas)).sum()
 
+    @lazy_property
+    def Tgas_cy_weighted(self):
+        if self.Ngas == 0:
+            return None
+        return (
+            self.gas_temperatures * (self.gas_compY.value / self.gas_compY.sum().value)
+        ).sum()
+
+    @lazy_property
+    def Tgas_cy_weighted_no_agn(self):
+        if self.Ngas == 0:
+            return None
+        return (
+            self.gas_temperatures[self.gas_no_agn]
+            * (
+                self.gas_compY[self.gas_no_agn].value
+                / self.gas_compY[self.gas_no_agn].sum().value
+            )
+        ).sum()
+
+    @lazy_property
+    def Tgas_cy_weighted_core_excision(self):
+        if self.Ngas == 0:
```

Comment:
This is the same problem again, and this one does generate a division by zero warning if I run on a L1000N0900/HYDRO_FIDUCIAL snapshot. There must be halos where all of the gas is in the excised core.

Reply:
I think this should be done for core-excision for sure indeed. For the others we could probably get away with it for the FLAMINGO cuts, but I'd be happy to code up and Ngas for all the selections

Reply:
I've added this in [3f8a13c](https://github.com/SWIFTSIM/SOAP/pull/5/commits/3f8a13c189d286e09f91f854f19bc8254d750818) it would be great if you could have a look

Reply:
We seem to have an infinite loop because gas_selection_core_excision_xray_temperature() and Ngas_core_excision_xray_temperature() call each other.

Reply:
Maybe we can remove all of the if statements from the gas_selection_* functions and just rely on the fact that gas properties are not calculated at all in DMO mode so the functions should never be called.

Reply:
I don't like returning selection=None when there are no particles anyway because None is a valid array slice which returns the full array. Probably better to let it raise an exception if the code tries to use a gas selection in a DMO run.

Reply:
Yeah I agree, the self-calling was a typo. I think it's only included on that one. But better to fix it completely.

Reply:
With that change it runs but still throws a few divide by zero warnings. They seem to come from the spectroscopic like temperature functions. I've tried to fix them in cdeb2b5.

Reply:
Yeah I mislabeled those. I didn't realise all the spectroscopic like temperatures used that mask.

Reply:
I also removed the None possibility for the three selections that aren't labeled as gas_selection in [41d67f2](https://github.com/SWIFTSIM/SOAP/pull/5/commits/41d67f2781383e0e721cc68b20bd675778c97a8a).